src/pip/_internal/vcs/__init__.py

```diff
@@ -25,6 +26,68 @@
 logger = logging.getLogger(__name__)
 
 
+class RevOptions(object):
+
+    """
+    Encapsulates a VCS-specific revision to install, along with any VCS
+    install options.
+
+    Instances of this class should be treated as if immutable.
+    """
+
+    def __init__(self, vcs, rev=None, extra_args=None):
+        """
+        Args:
+          vcs: a VersionControl object.
+          rev: the name of the revision to install.
+          extra_args: a list of extra options.
+        """
+        if extra_args is None:
+            extra_args = []
+
+        self.extra_args = extra_args
+        self.rev = rev
+        self.vcs = vcs
+
+    @property
+    def arg_rev(self):
+        if self.rev is None:
+            return self.vcs.default_arg_rev
+
+        return self.rev
+
+    def to_args(self, start_args, end_args=None):
```

Comment:
I'm not a fan:
I'd say `cmd_args = ['branch', '-q'] + rev_options.to_args() + [url, dest]` reads better.
An `__add__` might be overkill.

Reply:
That's a fair point. Thanks. I agree.

And thanks also for taking the time to review!